src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import Logger from '../logger/Logger';
+import Maybe from '../maybe/Maybe';
+import AsyncScheduler from '../scheduler/AsyncScheduler';
+import SimulcastLayers from '../simulcastlayers/SimulcastLayers';
+import SimulcastTransceiverController from '../transceivercontroller/SimulcastTransceiverController';
+import DefaultVideoAndEncodeParameter from '../videocaptureandencodeparameter/DefaultVideoCaptureAndEncodeParameter';
+import VideoStreamDescription from '../videostreamindex/VideoStreamDescription';
+import VideoStreamIndex from '../videostreamindex/VideoStreamIndex';
+import BitrateParameters from './BitrateParameters';
+import ConnectionMetrics from './ConnectionMetrics';
+import SimulcastUplinkObserver from './SimulcastUplinkObserver';
+import SimulcastUplinkPolicy from './SimulcastUplinkPolicy';
+
+const enum ActiveStreams {
+  kHiAndLow,
+  kMidAndLow,
+  kLow,
+}
+
+/**
+ * [[DefaultSimulcastUplinkPolicy]] determines capture and encode
+ *  parameters that reacts to estimated uplink bandwidth
+ */
+export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPolicy {
+  static readonly defaultUplinkBandwidthKbps: number = 1200;
+  static readonly startupDurationMs: number = 6000;
+  static readonly holdDownDurationMs: number = 4000;
+  static readonly defaultMaxFrameRate = 15;
+
+  // Current rough estimates where webrtc disables streams
+  static readonly kHiDisabledRate = 700;
+  static readonly kMidDisabledRate = 240;
+
+  private numParticipants: number = 0;
+  private optimalParameters: DefaultVideoAndEncodeParameter;
+  private parametersInEffect: DefaultVideoAndEncodeParameter;
+
+  private newQualityMap = new Map<string, RTCRtpEncodingParameters>();
+  private currentQualityMap = new Map<string, RTCRtpEncodingParameters>();
+  private newActiveStreams: ActiveStreams = ActiveStreams.kHiAndLow;
+  private currentActiveStreams: ActiveStreams = ActiveStreams.kHiAndLow;
+  private lastUplinkBandwidthKbps: number = DefaultSimulcastUplinkPolicy.defaultUplinkBandwidthKbps;
+
+  private startTimeMs: number = 0;
+  private lastUpdatedMs: number = Date.now();
+
+  private videoIndex: VideoStreamIndex | null = null;
+
+  private currLocalDescriptions: VideoStreamDescription[] = [];
+  private nextLocalDescriptions: VideoStreamDescription[] = [];
+
+  private activeSimulcastLayers: SimulcastLayers;
+  private observerQueue: Set<SimulcastUplinkObserver> = new Set<SimulcastUplinkObserver>();
+
+  constructor(private selfAttendeeId: string, private logger: Logger) {
+    this.optimalParameters = new DefaultVideoAndEncodeParameter(0, 0, 0, 0, true);
+    this.parametersInEffect = new DefaultVideoAndEncodeParameter(0, 0, 0, 0, true);
+
```

Comment:
Blank Line can be removed in a method definition.

Reply:
Done ✅